google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -154,32 +163,91 @@ def reset
 
         def close
           @thread.kill if @thread
-          @all_sessions.each(&:delete_session)
-          @all_sessions = []
-          @session_queue = []
-          @transaction_queue = []
+          # Unblock all waiting threads
+          @closed = true
+          @resource.broadcast
+          # Delete all sessions
+          @mutex.synchronize do
+            @all_sessions.map do |s|
+              Thread.new do
```

Comment:
We've been doing our best to avoid unnecessary external dependencies. It can add compatibility conflicts for users, as we've experienced in other Google Cloud libraries. But if its a hard requirement then we can't avoid it. I'll give it some thought over the weekend.